File: Sources/ComputerBar/SSHConfigService.swift
import Foundation

enum SSHConfigLoadError: LocalizedError {
    case missingConfig(URL)
    case unreadableConfig(URL, underlying: Error)
    case noHostsFound(URL)

    var errorDescription: String? {
        switch self {
        case let .missingConfig(url):
            return "Could not find SSH config at \(NSString(string: url.path).abbreviatingWithTildeInPath)."
        case let .unreadableConfig(url, underlying):
            return "Could not read \(NSString(string: url.path).abbreviatingWithTildeInPath): \(underlying.localizedDescription)"
        case let .noHostsFound(url):
            return "No concrete Host aliases were found in \(NSString(string: url.path).abbreviatingWithTildeInPath)."
        }
    }
}

struct SSHConfigService {
    let configURL: URL

    init(configURL: URL = FileManager.default.homeDirectoryForCurrentUser.appending(path: ".ssh/config")) {
        self.configURL = configURL
    }

    func loadHosts() async throws -> [SSHHost] {
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: configURL.path) else {
            throw SSHConfigLoadError.missingConfig(configURL)
        }

        let contents: String
        do {
            contents = try String(contentsOf: configURL, encoding: .utf8)
        } catch {
            throw SSHConfigLoadError.unreadableConfig(configURL, underlying: error)
        }

        let aliases = SSHConfigParser.parseAliases(from: contents)
        guard !aliases.isEmpty else {
            throw SSHConfigLoadError.noHostsFound(configURL)
        }

        var hosts: [SSHHost] = []
        for alias in aliases {
            hosts.append(await resolve(alias: alias))
        }

        return hosts
    }

    private func resolve(alias: String) async -> SSHHost {
        do {
            let output = try await ProcessRunner.run(
                executableURL: URL(fileURLWithPath: "/usr/bin/ssh"),
                arguments: ["-G", alias],
                timeout: 3
            )

            guard output.exitCode == 0 else {
                return SSHHost(alias: alias, hostName: alias, user: nil, port: nil)
            }

            return SSHConfigParser.parseResolvedHost(alias: alias, output: output.stdout)
        } catch {
            return SSHHost(alias: alias, hostName: alias, user: nil, port: nil)
        }
    }
}

enum SSHConfigParser {
    static func parseAliases(from contents: String) -> [String] {
        var aliases: [String] = []
        var seen = Set<String>()

        for rawLine in contents.split(whereSeparator: \.isNewline) {
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !line.isEmpty, !line.hasPrefix("#") else { continue }

            let parts = line.split(whereSeparator: \.isWhitespace)
            guard let first = parts.first, first.lowercased() == "host" else { continue }

            for token in parts.dropFirst() {
                let alias = String(token)
                if alias.hasPrefix("#") {
                    break
                }
                guard isConcreteAlias(alias) else { continue }
                if seen.insert(alias).inserted {
                    aliases.append(alias)
                }
            }
        }

        return aliases
    }

    static func parseResolvedHost(alias: String, output: String) -> SSHHost {
        var hostName: String?
        var user: String?
        var port: Int?

        for rawLine in output.split(whereSeparator: \.isNewline) {
            let parts = rawLine.split(maxSplits: 1, whereSeparator: \.isWhitespace)
            guard parts.count == 2 else { continue }

            let key = parts[0].lowercased()
            let value = parts[1].trimmingCharacters(in: .whitespacesAndNewlines)

            switch key {
            case "hostname":
                if hostName == nil, !value.isEmpty {
                    hostName = value
                }
            case "user":
                if user == nil, !value.isEmpty {
                    user = value
                }
            case "port":
                if port == nil {
                    port = Int(value)
                }
            default:
                continue
            }
        }

        return SSHHost(
            alias: alias,
            hostName: hostName ?? alias,
            user: user,
            port: port
        )
    }

    private static func isConcreteAlias(_ alias: String) -> Bool {
        guard !alias.isEmpty else { return false }
        guard !alias.hasPrefix("!") else { return false }
        guard !alias.contains("*"), !alias.contains("?") else { return false }
        return true
    }
}

File: Windows/src/ComputerBar.Core/SshConfigService.cs
namespace ComputerBar.Core;

public enum SshConfigLoadErrorKind
{
    MissingConfig,
    UnreadableConfig,
    NoHostsFound
}

public sealed class SshConfigLoadException(
    SshConfigLoadErrorKind kind,
    string path,
    Exception? innerException = null)
    : Exception(BuildMessage(kind, path, innerException), innerException)
{
    public SshConfigLoadErrorKind Kind { get; } = kind;
    public string PathValue { get; } = path;

    private static string BuildMessage(SshConfigLoadErrorKind kind, string path, Exception? innerException) => kind switch
    {
        SshConfigLoadErrorKind.MissingConfig => $"Could not find SSH config at {path}.",
        SshConfigLoadErrorKind.UnreadableConfig => $"Could not read {path}: {innerException?.Message ?? "unknown error"}",
        SshConfigLoadErrorKind.NoHostsFound => $"No concrete Host aliases were found in {path}.",
        _ => $"Could not load SSH config at {path}."
    };
}

public sealed class SshConfigService(string? configFile = null) : IHostConfigService
{
    public string ConfigFile { get; } = configFile ?? ComputerBarPaths.DefaultSshConfigFile;

    public async Task<IReadOnlyList<ComputerHost>> LoadHostsAsync(CancellationToken cancellationToken = default)
    {
        if (!File.Exists(ConfigFile))
        {
            throw new SshConfigLoadException(SshConfigLoadErrorKind.MissingConfig, ConfigFile);
        }

        string contents;
        try
        {
            contents = await File.ReadAllTextAsync(ConfigFile, cancellationToken);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            throw new SshConfigLoadException(SshConfigLoadErrorKind.UnreadableConfig, ConfigFile, ex);
        }

        var hosts = SshConfigParser.ParseHosts(contents);
        if (hosts.Count == 0)
        {
            throw new SshConfigLoadException(SshConfigLoadErrorKind.NoHostsFound, ConfigFile);
        }

        return hosts;
    }
}

public static class SshConfigParser
{
    public static IReadOnlyList<string> ParseAliases(string contents) =>
        ParseHosts(contents).Select(host => host.Alias).ToArray();

    public static IReadOnlyList<ComputerHost> ParseHosts(string contents)
    {
        var hosts = new List<ComputerHost>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var aliases = new List<string>();
        string? hostName = null;
        string? user = null;
        int? port = null;

        void CommitBlock()
        {
            foreach (var alias in aliases)
            {
                if (!seen.Add(alias))
                {
                    continue;
                }

                hosts.Add(new ComputerHost(alias, hostName ?? alias, user, port));
            }
        }

        foreach (var rawLine in contents.Split(["\r\n", "\n"], StringSplitOptions.None))
        {
            var line = TrimInlineComment(rawLine).Trim();
            if (line.Length == 0)
            {
                continue;
            }

            var parts = line.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length == 0)
            {
                continue;
            }

            var key = parts[0];
            if (key.Equals("Host", StringComparison.OrdinalIgnoreCase))
            {
                CommitBlock();
                aliases = parts.Skip(1)
                    .TakeWhile(token => !token.StartsWith('#'))
                    .Where(IsConcreteAlias)
                    .Distinct(StringComparer.OrdinalIgnoreCase)
                    .ToList();
                hostName = null;
                user = null;
                port = null;
                continue;
            }

            if (aliases.Count == 0 || parts.Length < 2)
            {
                continue;
            }

            var value = string.Join(" ", parts.Skip(1)).Trim();
            switch (key.ToLowerInvariant())
            {
                case "hostname":
                    if (hostName is null && value.Length > 0)
                    {
                        hostName = value;
                    }
                    break;
                case "user":
                    if (user is null && value.Length > 0)
                    {
                        user = value;
                    }
                    break;
                case "port":
                    if (port is null && int.TryParse(value, out var parsedPort))
                    {
                        port = parsedPort;
                    }
                    break;
            }
        }

        CommitBlock();
        return hosts;
    }

    private static string TrimInlineComment(string line)
    {
        for (var index = 0; index < line.Length; index++)
        {
            if (line[index] == '#' && (index == 0 || char.IsWhiteSpace(line[index - 1])))
            {
                return line[..index];
            }
        }

        return line;
    }

    private static bool IsConcreteAlias(string alias)
    {
        if (string.IsNullOrWhiteSpace(alias))
        {
            return false;
        }

        return !alias.StartsWith('!')
            && !alias.Contains('*')
            && !alias.Contains('?');
    }
}
